server/src/main/java/org/opensearch/gateway/GatewayAllocator.java

```diff
@@ -168,6 +246,169 @@ public void allocateUnassigned(
+        }
+    }
+
+    // visible for testing
+    protected Set<String> createAndUpdateBatches(RoutingAllocation allocation, boolean primary) {
+        Set<String> batchesToBeAssigned = new HashSet<>();
+        RoutingNodes.UnassignedShards unassigned = allocation.routingNodes().unassigned();
+        ConcurrentMap<String, ShardsBatch> currentBatches = primary ? batchIdToStartedShardBatch : batchIdToStoreShardBatch;
+        // get all batched shards
+        Set<ShardId> currentBatchedShards = currentBatches.values()
+            .stream()
+            .map(ShardsBatch::getBatchedShards)
+            .flatMap(Set::stream)
+            .collect(Collectors.toSet());
+
+        Set<ShardRouting> shardsToBatch = Sets.newHashSet();
+        // add all unassigned shards to the batch if they are not already in a batch
+        unassigned.forEach(shardRouting -> {
+            if ((currentBatchedShards.contains(shardRouting.shardId()) == false) && (shardRouting.primary() == primary)) {
+                assert shardRouting.unassigned();
+                shardsToBatch.add(shardRouting);
+            }
+            // if shard is already batched update to latest shardRouting information in the batches
+            else if (shardRouting.primary() == primary) {
+                String batchId = getBatchId(shardRouting, shardRouting.primary());
+                batchesToBeAssigned.add(batchId);
+                currentBatches.get(batchId).batchInfo.get(shardRouting.shardId()).setShardRouting(shardRouting);
+            }
+        });
+        Iterator<ShardRouting> iterator = shardsToBatch.iterator();
+        assert maxBatchSize > 0 : "Shards batch size must be greater than 0";
+
+        long batchSize = maxBatchSize;
+        Map<ShardId, ShardEntry> addToCurrentBatch = new HashMap<>();
+        while (iterator.hasNext()) {
+            ShardRouting currentShard = iterator.next();
+            if (batchSize > 0) {
+                ShardEntry sharEntry = new ShardEntry(
+                    IndexMetadata.INDEX_DATA_PATH_SETTING.get(allocation.metadata().index(currentShard.index()).getSettings()),
+                    currentShard
+                );
+                addToCurrentBatch.put(currentShard.shardId(), sharEntry);
+                batchSize--;
+                iterator.remove();
+            }
+            // add to batch if batch size full or last shard in unassigned list
+            if (batchSize == 0 || iterator.hasNext() == false) {
+                String batchUUId = UUIDs.base64UUID();
+                ShardsBatch shardsBatch = new ShardsBatch(batchUUId, addToCurrentBatch, primary);
+                // add the batch to list of current batches
+                addBatch(shardsBatch, primary);
+                batchesToBeAssigned.add(batchUUId);
+                addToCurrentBatch.clear();
+                batchSize = maxBatchSize;
+            }
+        }
```

Comment:
Extract out this logic into a separate method. Let us not have a 100 line long method unless explicitly required for synchronization and such requirement.